src/core/ext/transport/chttp2/transport/hpack_encoder.cc

```diff
@@ -431,13 +431,27 @@ void HPackCompressor::Framer::Encode(HttpSchemeMetadata,
 }
 
 void HPackCompressor::Framer::Encode(GrpcTraceBinMetadata, const Slice& slice) {
+  const size_t table_entry_size = GrpcTraceBinMetadata::key().size() +
+                                  slice.size() +
+                                  hpack_constants::kEntryOverhead;
+  if (table_entry_size > HPackEncoderTable::MaxEntrySize()) {
+    EmitLitHdrWithBinaryStringKeyNotIdx(
+        Slice::FromStaticString(GrpcTraceBinMetadata::key()), slice.Ref());
+  }
   EncodeIndexedKeyWithBinaryValue(&compressor_->grpc_trace_bin_index_,
-                                  "grpc-trace-bin", slice.Ref());
+                                  GrpcTraceBinMetadata::key(), slice.Ref());
 }
 
 void HPackCompressor::Framer::Encode(GrpcTagsBinMetadata, const Slice& slice) {
+  const size_t table_entry_size = GrpcTagsBinMetadata::key().size() +
+                                  slice.size() +
```

Comment:
Seems like a `HPackCompressor::Framer::EncodeRepeatingSliceValue(absl::string_view key, const Slice& slice)` would be useful here, so we could reduce these Encode methods to one liners.

(repeating or some such word to indicate that hpack compression should be employed if possible)

It might be worth adding a third argument of `max_compress_length` or some such so that we can control the maximum hpack table usage per header field - some fields we know are more important than others and we'd be more willing to take memory for them.